storage/simulation/range.go

```diff
@@ -116,10 +128,59 @@ func (r *Range) splitRange(originalRange *Range) {
 	stores := originalRange.getStores()
 	r.Lock()
 	defer r.Unlock()
-	r.desc.Replicas = desc.Replicas
+	r.desc.Replicas = append([]proto.Replica(nil), desc.Replicas...)
 	r.stores = stores
 }
 
+// prepareActions walks through each replica and determines if any action is
+// required using the allocator.
+func (r *Range) prepareActions() {
+	r.Lock()
+	defer r.Unlock()
+	r.nextActions = make(map[proto.StoreID]nextAction)
+	for storeID := range r.stores {
+		action, priority := r.allocator.ComputeAction(r.zone, &r.desc)
+		var rebalance bool
+		if action == storage.AANoop {
+			rebalance = r.allocator.ShouldRebalance(storeID)
+			priority = 0
+		}
+		r.nextActions[storeID] = nextAction{
+			action:    action,
+			priority:  priority,
+			rebalance: rebalance,
+		}
+	}
+}
+
+// getNextAction returns a single action based on the priorities of the each
+// action stored in nextActions.
+func (r *Range) getNextAction() nextAction {
+	r.RLock()
+	defer r.RUnlock()
+	var topAction nextAction
+	// TODO(bram): This is random. Might want to make it deterministic for
+	// repeatability.
+	for _, nextAction := range r.nextActions {
+		if nextAction.priority > topAction.priority {
+			topAction = nextAction
+		}
+	}
+	return topAction
+}
+
+// getAllocateTarget calls allocateTarget for the range and returns the top
+// target store.
+func (r *Range) getAllocateTarget() proto.StoreID {
+	r.RLock()
```

Comment:
There's nothing wrong with acquiring a read lock twice.  
But,  i'm going to remove all of the mutexes (and gets) in an upcoming PR.  The intention was to fire all actions at the same time to better emulate a real system, but that's something that I can add on later.